File: scripts/ib_realtime_server.js
#!/usr/bin/env node

/**
 * Interactive Brokers Real-Time Price Server (Node.js)
 *
 * This is a direct replacement for the Python websocket server.
 */

import process from "node:process";
import { WebSocketServer } from "ws";
import IB from "ib";

const DEFAULT_WS_PORT = 8765;
const DEFAULT_IB_HOST = "127.0.0.1";
const DEFAULT_IB_PORT = 4001;
const RECONNECT_MS = 5000;
const SNAPSHOT_TIMEOUT_MS = 5000;

function parseArgs(argv) {
  const args = {
    port: DEFAULT_WS_PORT,
    ibHost: DEFAULT_IB_HOST,
    ibPort: DEFAULT_IB_PORT,
  };

  for (let i = 0; i < argv.length; i += 1) {
    const arg = argv[i];
    if (arg === "--port") {
      const value = Number.parseInt(argv[i + 1], 10);
      if (Number.isInteger(value)) {
        args.port = value;
      }
      i += 1;
      continue;
    }
    if (arg === "--ib-host") {
      args.ibHost = argv[i + 1] ?? DEFAULT_IB_HOST;
      i += 1;
      continue;
    }
    if (arg === "--ib-port") {
      const value = Number.parseInt(argv[i + 1], 10);
      if (Number.isInteger(value)) {
        args.ibPort = value;
      }
      i += 1;
    }
  }

  return args;
}

function normalizeSymbols(raw) {
  return raw
    .map((symbol) => String(symbol).trim().toUpperCase())
    .filter((symbol) => symbol.length > 0);
}

function normalizeNumber(value) {
  if (typeof value !== "number" || !Number.isFinite(value)) {
    return null;
  }
  return value;
}

function nowIso() {
  return new Date().toISOString();
}

function createPriceData(symbol) {
  return {
    symbol,
    last: null,
    lastIsCalculated: false,
    bid: null,
    ask: null,
    bidSize: null,
    askSize: null,
    volume: null,
    high: null,
    low: null,
    open: null,
    close: null,
    timestamp: nowIso(),
  };
}

function updateDerivedLast(data) {
  if (data.last == null && data.bid != null && data.ask != null) {
    const midpoint = (data.bid + data.ask) / 2;
    data.last = Number.isFinite(midpoint) ? Number(midpoint.toFixed(4)) : null;
    data.lastIsCalculated = true;
  }
}

function updatePriceFromTickPrice(data, tickType, value) {
  const { TICK_TYPE } = IB;
  switch (tickType) {
    case TICK_TYPE.BID:
      data.bid = normalizeNumber(value);
      data.lastIsCalculated = false;
      break;
    case TICK_TYPE.ASK:
      data.ask = normalizeNumber(value);
      data.lastIsCalculated = false;
      break;
    case TICK_TYPE.LAST:
      data.last = normalizeNumber(value);
      data.lastIsCalculated = false;
      break;
    case TICK_TYPE.HIGH:
      data.high = normalizeNumber(value);
      break;
    case TICK_TYPE.LOW:
      data.low = normalizeNumber(value);
      break;
    case TICK_TYPE.OPEN:
      data.open = normalizeNumber(value);
      break;
    case TICK_TYPE.CLOSE:
      data.close = normalizeNumber(value);
      break;
    case TICK_TYPE.VOLUME:
      data.volume = normalizeNumber(value);
      break;
    default:
      break;
  }

  if (data.last == null) {
    updateDerivedLast(data);
  }
  data.timestamp = nowIso();
}

function updatePriceFromTickSize(data, sizeType, value) {
  const { TICK_TYPE } = IB;
  switch (sizeType) {
    case TICK_TYPE.BID_SIZE:
      data.bidSize = normalizeNumber(value);
      break;
    case TICK_TYPE.ASK_SIZE:
      data.askSize = normalizeNumber(value);
      break;
    case TICK_TYPE.VOLUME:
      data.volume = normalizeNumber(value);
      break;
    case TICK_TYPE.LAST_SIZE:
      break;
    default:
      break;
  }

  data.timestamp = nowIso();
}

function parseActionMessage(raw) {
  if (typeof raw !== "object" || raw === null || Array.isArray(raw)) return null;

  const payload = raw;
  if (typeof payload.action !== "string") {
    return null;
  }

  const action = payload.action.trim().toLowerCase();
  if (!action) {
    return null;
  }

  const symbols = Array.isArray(payload.symbols) ? normalizeSymbols(payload.symbols) : [];
  return { action, symbols };
}

const cli = parseArgs(process.argv.slice(2));
const wsUrl = `ws://0.0.0.0:${cli.port}`;

const ib = new IB({
  host: cli.ibHost,
  port: cli.ibPort,
  clientId: 100,
});

const wss = new WebSocketServer({ host: "0.0.0.0", port: cli.port });

const clients = new Set();
const symbolSubscribers = new Map();
const clientSymbols = new Map();
const symbolStates = new Map();
const requestIdToSymbol = new Map();
const snapshotRequests = new Map();

let ibConnected = false;
let shuttingDown = false;
let reconnectTimer = null;
let nextRequestId = 1;
let statusBroadcastTick = null;

function sendMessage(client, payload) {
  try {
    if (client.readyState === client.OPEN) {
      client.send(JSON.stringify(payload));
    }
  } catch {
    // Ignore send failures.
  }
}

function sendToSymbolSubscribers(symbol, payload) {
  const subscribers = symbolSubscribers.get(symbol);
  if (!subscribers || subscribers.size === 0) return;
  for (const client of subscribers) {
    sendMessage(client, payload);
  }
}

function sendStatus(client) {
  const subscriptions = Array.from(symbolSubscribers.keys()).filter((symbol) => symbolSubscribers.get(symbol)?.size);
  sendMessage(client, {
    type: "status",
    ib_connected: ibConnected,
    subscriptions,
  });
}

function broadcastStatus() {
  for (const client of clients) {
    sendStatus(client);
  }
}

function clearSnapshot(requestId) {
  const req = snapshotRequests.get(requestId);
  if (!req) return;
  clearTimeout(req.timer);
  snapshotRequests.delete(requestId);
  requestIdToSymbol.delete(requestId);
}

function completeSnapshot(symbol, requestId) {
  const req = snapshotRequests.get(requestId);
  if (!req) return;
  sendMessage(req.client, {
    type: "snapshot",
    symbol,
    data: req.data,
  });
  clearSnapshot(requestId);
  try {
    ib.cancelMktData(requestId);
  } catch {
    // Ignore cleanup failures.
  }
}

function startLiveSubscription(symbol) {
  if (!ibConnected) return;

  const existing = symbolStates.get(symbol);
  const nextTickerId = nextRequestId += 1;
  const contract = ib.contract.stock(symbol, "SMART", "USD");
  const state = existing ?? {
    tickerId: null,
    contract,
    data: createPriceData(symbol),
  };

  if (state.tickerId != null) {
    try {
      ib.cancelMktData(state.tickerId);
    } catch {
      // Ignore.
    }
    requestIdToSymbol.delete(state.tickerId);
  }

  try {
    ib.reqMktData(nextTickerId, contract, "233", false, false);
    state.tickerId = nextTickerId;
    state.contract = contract;
    state.data.timestamp = nowIso();
    symbolStates.set(symbol, state);
    requestIdToSymbol.set(nextTickerId, symbol);
  } catch (error) {
    console.error(`Failed to subscribe symbol ${symbol}:`, error);
  }
}

function stopLiveSubscription(symbol) {
  const state = symbolStates.get(symbol);
  if (!state || state.tickerId == null) return;
  try {
    ib.cancelMktData(state.tickerId);
  } catch {
    // Ignore.
  }
  requestIdToSymbol.delete(state.tickerId);
  symbolStates.delete(symbol);
}

function cleanupSymbolStateForReconnect() {
  for (const state of symbolStates.values()) {
    if (state.tickerId != null) {
      try {
        ib.cancelMktData(state.tickerId);
      } catch {
        // Ignore.
      }
      requestIdToSymbol.delete(state.tickerId);
      state.tickerId = null;
    }
  }
}

function subscribeClientToSymbol(client, symbol) {
  let subscribers = symbolSubscribers.get(symbol);
  if (!subscribers) {
    subscribers = new Set();
    symbolSubscribers.set(symbol, subscribers);
  }
  subscribers.add(client);

  let clientSet = clientSymbols.get(client);
  if (!clientSet) {
    clientSet = new Set();
    clientSymbols.set(client, clientSet);
  }
  clientSet.add(symbol);
}

function unsubscribeClientFromSymbol(client, symbol) {
  const subscribers = symbolSubscribers.get(symbol);
  let unsubscribed = false;

  if (subscribers) {
    subscribers.delete(client);
    if (subscribers.size === 0) {
      symbolSubscribers.delete(symbol);
      stopLiveSubscription(symbol);
      unsubscribed = true;
    } else {
      unsubscribed = true;
    }
  }

  const clientSet = clientSymbols.get(client);
  if (clientSet) {
    clientSet.delete(symbol);
  }

  return unsubscribed;
}

function disconnectClient(client) {
  const clientSet = clientSymbols.get(client);
  if (!clientSet) {
    return;
  }

  for (const symbol of clientSet) {
    const subscribers = symbolSubscribers.get(symbol);
    if (!subscribers) continue;

    subscribers.delete(client);
    if (subscribers.size === 0) {
      symbolSubscribers.delete(symbol);
      stopLiveSubscription(symbol);
    }
  }

  clientSymbols.delete(client);
}

function sendSubscribedConfirmation(client, symbols) {
  sendMessage(client, {
    type: "subscribed",
    symbols,
  });
}

function sendUnsubscribedConfirmation(client, symbols) {
  sendMessage(client, {
    type: "unsubscribed",
    symbols,
  });
}

async function handleSnapshotRequest(client, symbols) {
  for (const symbol of symbols) {
    if (!ibConnected) {
      sendMessage(client, {
        type: "error",
        message: "IB not connected",
      });
      continue;
    }

    const requestId = nextRequestId += 1;
    const contract = ib.contract.stock(symbol, "SMART", "USD");
    const requestState = {
      symbol,
      client,
      timer: setTimeout(() => {
        sendMessage(client, {
          type: "error",
          message: `Timeout waiting for snapshot: ${symbol}`,
        });
        clearSnapshot(requestId);
        try {
          ib.cancelMktData(requestId);
        } catch {
          // Ignore.
        }
      }, SNAPSHOT_TIMEOUT_MS),
      data: createPriceData(symbol),
    };

    snapshotRequests.set(requestId, requestState);
    requestIdToSymbol.set(requestId, symbol);

    try {
      ib.reqMktData(requestId, contract, "233", true, false);
    } catch (error) {
      clearSnapshot(requestId);
      try {
        ib.cancelMktData(requestId);
      } catch {
        // Ignore.
      }
      sendMessage(client, {
        type: "error",
        message: `Failed to request snapshot for ${symbol}: ${String(error)}`,
      });
    }
  }
}

function hydrateAndBroadcast(symbol) {
  const state = symbolStates.get(symbol);
  if (!state) return;
  sendToSymbolSubscribers(symbol, {
    type: "price",
    symbol,
    data: state.data,
  });
}

function onTickPrice(tickerId, tickType, price) {
  const symbol = requestIdToSymbol.get(tickerId);
  const liveState = symbol ? symbolStates.get(symbol) : null;
  const snapshotState = snapshotRequests.get(tickerId);

  if (liveState) {
    updatePriceFromTickPrice(liveState.data, tickType, price);
    hydrateAndBroadcast(symbol);
  }
  if (snapshotState) {
    updatePriceFromTickPrice(snapshotState.data, tickType, price);
  }
}

function onTickSize(tickerId, sizeType, size) {
  const symbol = requestIdToSymbol.get(tickerId);
  const liveState = symbol ? symbolStates.get(symbol) : null;
  const snapshotState = snapshotRequests.get(tickerId);

  if (liveState) {
    updatePriceFromTickSize(liveState.data, sizeType, size);
  }
  if (snapshotState) {
    updatePriceFromTickSize(snapshotState.data, sizeType, size);
  }
}

function onTickSnapshotEnd(tickerId) {
  const symbol = requestIdToSymbol.get(tickerId);
  if (!symbol) return;
  completeSnapshot(symbol, tickerId);
}

function restoreSubscriptions() {
  const symbols = [...symbolSubscribers.keys()];
  for (const symbol of symbols) {
    startLiveSubscription(symbol);
    const state = symbolStates.get(symbol);
    if (state) {
      sendToSymbolSubscribers(symbol, {
        type: "price",
        symbol,
        data: state.data,
      });
    }
  }
}

async function handleClientMessage(client, data) {
  const message = parseActionMessage(data);
  if (!message) {
    sendMessage(client, { type: "error", message: "Invalid JSON" });
    return;
  }

  const symbols = message.symbols;
  switch (message.action) {
    case "subscribe": {
      const subscribed = [];
      for (const symbol of symbols) {
        subscribeClientToSymbol(client, symbol);
        if (ibConnected) {
          startLiveSubscription(symbol);
          const state = symbolStates.get(symbol);
          if (state) {
            sendMessage(client, {
              type: "price",
              symbol,
              data: state.data,
            });
          }
          subscribed.push(symbol);
        }
      }
      sendSubscribedConfirmation(client, subscribed);
      return;
    }
    case "unsubscribe": {
      const unsubscribed = [];
      for (const symbol of symbols) {
        if (unsubscribeClientFromSymbol(client, symbol)) {
          unsubscribed.push(symbol);
        }
      }
      sendUnsubscribedConfirmation(client, unsubscribed);
      return;
    }
    case "snapshot": {
      await handleSnapshotRequest(client, symbols);
      return;
    }
    case "ping": {
      sendMessage(client, { type: "pong" });
      return;
    }
    default: {
      sendMessage(client, {
        type: "error",
        message: `Unknown action: ${message.action}`,
      });
    }
  }
}

function scheduleReconnect() {
  if (shuttingDown || reconnectTimer) return;
  reconnectTimer = setTimeout(() => {
    reconnectTimer = null;
    console.log(`Attempting IB reconnect to ${cli.ibHost}:${cli.ibPort}...`);
    try {
      ib.disconnect();
    } catch {
      // Ignore.
    }
    try {
      ib.connect();
    } catch {
      // Ignore.
      ibConnected = false;
      broadcastStatus();
      scheduleReconnect();
    }
  }, RECONNECT_MS);
}

ib.on("connected", () => {
  ibConnected = true;
  console.log("IB connected");
  reconnectTimer = null;
  cleanupSymbolStateForReconnect();
  restoreSubscriptions();
  broadcastStatus();
});

ib.on("disconnected", () => {
  if (ibConnected) {
    console.log("IB disconnected");
  }
  ibConnected = false;
  broadcastStatus();
  scheduleReconnect();
});

ib.on("error", (error) => {
  console.error("IB error:", error);
  broadcastStatus();
});

ib.on("tickPrice", (tickerId, tickType, price) => {
  onTickPrice(tickerId, tickType, price);
});

ib.on("tickSize", (tickerId, sizeType, size) => {
  onTickSize(tickerId, sizeType, size);
});

ib.on("tickSnapshotEnd", (tickerId) => {
  onTickSnapshotEnd(tickerId);
});

wss.on("connection", (client) => {
  clients.add(client);
  sendStatus(client);

  client.on("message", (raw) => {
    const payload = (() => {
      if (typeof raw === "string") return raw;
      if (raw instanceof Buffer) return raw.toString("utf8");
      if (raw instanceof ArrayBuffer) return new TextDecoder().decode(raw);
      return "";
    })();

    if (!payload) return;

    try {
      const data = JSON.parse(payload);
      void handleClientMessage(client, data);
    } catch {
      sendMessage(client, { type: "error", message: "Invalid JSON" });
    }
  });

  client.on("close", () => {
    disconnectClient(client);
    clients.delete(client);
  });

  client.on("error", () => {
    disconnectClient(client);
    clients.delete(client);
  });
});

ib.connect();

statusBroadcastTick = setInterval(() => {
  if (ibConnected) return;
  for (const client of clients) {
    sendStatus(client);
  }
}, 5000);

process.on("SIGINT", () => {
  if (shuttingDown) process.exit(0);
  shuttingDown = true;
  if (reconnectTimer) {
    clearTimeout(reconnectTimer);
  }
  if (statusBroadcastTick) {
    clearInterval(statusBroadcastTick);
  }
  for (const client of clients) {
    try {
      client.close();
    } catch {
      // Ignore.
    }
  }
  for (const [requestId] of snapshotRequests) {
    clearSnapshot(requestId);
  }
  for (const state of symbolStates.values()) {
    if (state.tickerId != null) {
      try {
        ib.cancelMktData(state.tickerId);
      } catch {
        // Ignore.
      }
    }
  }
  try {
    wss.close();
    ib.disconnect();
  } catch {
    // Ignore.
  }
  process.exit(0);
});

process.on("SIGTERM", () => {
  process.emit("SIGINT");
});

console.log(`IB realtime server listening on ${wsUrl}`);
console.log(`IB target ${cli.ibHost}:${cli.ibPort}`);

File: scripts/ib_realtime_server.py
#!/usr/bin/env python3
"""
Interactive Brokers Real-Time Price Server

A standalone WebSocket server that streams real-time prices from IB.
Separate from ib_sync.py which handles periodic portfolio syncing.

Usage:
  python3 scripts/ib_realtime_server.py
  python3 scripts/ib_realtime_server.py --port 8765 --ib-port 4001

Requirements:
  pip install ib_insync websockets

Protocol:
  Client -> Server:
    {"action": "subscribe", "symbols": ["AAPL", "MSFT"]}
    {"action": "unsubscribe", "symbols": ["AAPL"]}
    {"action": "snapshot", "symbols": ["NVDA"]}
    {"action": "ping"}
  
  Server -> Client:
    {"type": "price", "symbol": "AAPL", "data": {...}}
    {"type": "subscribed", "symbols": ["AAPL", "MSFT"]}
    {"type": "unsubscribed", "symbols": ["AAPL"]}
    {"type": "snapshot", "symbol": "NVDA", "data": {...}}
    {"type": "error", "message": "..."}
    {"type": "pong"}
    {"type": "status", "ib_connected": true, "subscriptions": [...]}
"""

import argparse
import asyncio
import json
import logging
import signal
import sys
from datetime import datetime
from typing import Dict, Set, Optional, Any

try:
    from ib_insync import IB, Stock, Option, util
    HAS_IB = True
except ImportError:
    HAS_IB = False
    print("Warning: ib_insync not installed. Run: pip install ib_insync", file=sys.stderr)

try:
    import websockets
    from websockets.server import WebSocketServerProtocol
    HAS_WS = True
except ImportError:
    HAS_WS = False
    print("Warning: websockets not installed. Run: pip install websockets", file=sys.stderr)


# Configure logging
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s [%(levelname)s] %(message)s',
    datefmt='%H:%M:%S'
)
logger = logging.getLogger(__name__)


class IBRealtimeServer:
    """WebSocket server that streams real-time IB prices."""
    
    def __init__(self, ws_port: int = 8765, ib_host: str = "127.0.0.1", ib_port: int = 4001):
        self.ws_port = ws_port
        self.ib_host = ib_host
        self.ib_port = ib_port
        self.ib: Optional[IB] = None
        self.clients: Set[WebSocketServerProtocol] = set()
        self.subscriptions: Dict[str, Set[WebSocketServerProtocol]] = {}  # symbol -> clients
        self.tickers: Dict[str, Any] = {}  # symbol -> IB ticker
        self.contracts: Dict[str, Any] = {}  # symbol -> IB contract
        self.running = False
    
    async def connect_ib(self) -> bool:
        """Connect to IB Gateway/TWS."""
        if not HAS_IB:
            logger.error("ib_insync not installed")
            return False
        
        try:
            self.ib = IB()
            await self.ib.connectAsync(
                self.ib_host, 
                self.ib_port, 
                clientId=100,  # Use dedicated client ID for realtime server
                timeout=10
            )
            logger.info(f"✓ Connected to IB on {self.ib_host}:{self.ib_port}")
            
            # Set up disconnect handler
            self.ib.disconnectedEvent += self._on_ib_disconnect
            
            return True
        except Exception as e:
            logger.error(f"✗ Failed to connect to IB: {e}")
            return False
    
    def _on_ib_disconnect(self):
        """Handle IB disconnection."""
        logger.warning("IB connection lost, attempting reconnect...")
        asyncio.create_task(self._reconnect_ib())
    
    async def _reconnect_ib(self):
        """Attempt to reconnect to IB."""
        await asyncio.sleep(5)
        if await self.connect_ib():
            # Resubscribe to all symbols
            symbols = list(self.subscriptions.keys())
            for symbol in symbols:
                await self._subscribe_symbol(symbol)
    
    async def _subscribe_symbol(self, symbol: str) -> bool:
        """Subscribe to market data for a symbol."""
        if not self.ib or not self.ib.isConnected():
            return False
        
        if symbol in self.tickers:
            return True  # Already subscribed
        
        try:
            # Create contract
            contract = Stock(symbol, 'SMART', 'USD')
            await self.ib.qualifyContractsAsync(contract)
            
            # Request streaming market data
            ticker = self.ib.reqMktData(contract, '233', False, False)
            
            # Set up update callback
            ticker.updateEvent += lambda t: asyncio.create_task(
                self._on_ticker_update(symbol, t)
            )
            
            self.tickers[symbol] = ticker
            self.contracts[symbol] = contract
            
            logger.info(f"Subscribed to {symbol}")
            return True
            
        except Exception as e:
            logger.error(f"Failed to subscribe to {symbol}: {e}")
            return False
    
    async def _unsubscribe_symbol(self, symbol: str):
        """Unsubscribe from market data for a symbol."""
        if symbol in self.tickers and self.ib and self.ib.isConnected():
            try:
                self.ib.cancelMktData(self.contracts[symbol])
            except:
                pass
            del self.tickers[symbol]
            del self.contracts[symbol]
            logger.info(f"Unsubscribed from {symbol}")
    
    async def _on_ticker_update(self, symbol: str, ticker: Any):
        """Handle ticker update from IB."""
        if symbol not in self.subscriptions:
            return
        
        # Build price update message
        data = self._ticker_to_dict(symbol, ticker)
        message = json.dumps({"type": "price", "symbol": symbol, "data": data})
        
        # Broadcast to all subscribed clients
        clients = self.subscriptions.get(symbol, set())
        if clients:
            await asyncio.gather(
                *[self._safe_send(client, message) for client in clients],
                return_exceptions=True
            )
    
    def _ticker_to_dict(self, symbol: str, ticker: Any) -> dict:
        """Convert IB ticker to dictionary."""
        last = ticker.last if ticker.last == ticker.last else None  # NaN check
        bid = ticker.bid if ticker.bid == ticker.bid else None
        ask = ticker.ask if ticker.ask == ticker.ask else None
        last_is_calculated = False
        if last is None and bid is not None and ask is not None:
            last = round((bid + ask) / 2, 4)
            last_is_calculated = True

        return {
            "symbol": symbol,
            "last": last,
            "lastIsCalculated": last_is_calculated,
            "bid": bid,
            "ask": ask,
            "bidSize": ticker.bidSize if ticker.bidSize == ticker.bidSize else None,
            "askSize": ticker.askSize if ticker.askSize == ticker.askSize else None,
            "volume": ticker.volume if ticker.volume == ticker.volume else None,
            "high": ticker.high if ticker.high == ticker.high else None,
            "low": ticker.low if ticker.low == ticker.low else None,
            "open": ticker.open if ticker.open == ticker.open else None,
            "close": ticker.close if ticker.close == ticker.close else None,
            "timestamp": datetime.now().isoformat()
        }
    
    async def _safe_send(self, client: WebSocketServerProtocol, message: str):
        """Safely send message to client, handling disconnections."""
        try:
            await client.send(message)
        except websockets.exceptions.ConnectionClosed:
            await self._remove_client(client)
    
    async def _remove_client(self, client: WebSocketServerProtocol):
        """Remove a client and clean up its subscriptions."""
        self.clients.discard(client)
        
        # Remove from all subscriptions
        empty_symbols = []
        for symbol, clients in self.subscriptions.items():
            clients.discard(client)
            if not clients:
                empty_symbols.append(symbol)
        
        # Unsubscribe from symbols with no clients
        for symbol in empty_symbols:
            del self.subscriptions[symbol]
            await self._unsubscribe_symbol(symbol)
    
    async def handle_client(self, websocket: WebSocketServerProtocol):
        """Handle a WebSocket client connection."""
        self.clients.add(websocket)
        client_id = id(websocket)
        logger.info(f"Client {client_id} connected ({len(self.clients)} total)")
        
        # Send initial status
        await self._safe_send(websocket, json.dumps({
            "type": "status",
            "ib_connected": self.ib.isConnected() if self.ib else False,
            "subscriptions": list(self.subscriptions.keys())
        }))
        
        try:
            async for message in websocket:
                await self._handle_message(websocket, message)
        except websockets.exceptions.ConnectionClosed:
            pass
        finally:
            await self._remove_client(websocket)
            logger.info(f"Client {client_id} disconnected ({len(self.clients)} total)")
    
    async def _handle_message(self, client: WebSocketServerProtocol, message: str):
        """Handle incoming message from client."""
        try:
            data = json.loads(message)
            action = data.get("action")
            
            if action == "subscribe":
                symbols = data.get("symbols", [])
                await self._handle_subscribe(client, symbols)
            
            elif action == "unsubscribe":
                symbols = data.get("symbols", [])
                await self._handle_unsubscribe(client, symbols)
            
            elif action == "snapshot":
                symbols = data.get("symbols", [])
                await self._handle_snapshot(client, symbols)
            
            elif action == "ping":
                await self._safe_send(client, json.dumps({"type": "pong"}))
            
            else:
                await self._safe_send(client, json.dumps({
                    "type": "error",
                    "message": f"Unknown action: {action}"
                }))
                
        except json.JSONDecodeError:
            await self._safe_send(client, json.dumps({
                "type": "error",
                "message": "Invalid JSON"
            }))
    
    async def _handle_subscribe(self, client: WebSocketServerProtocol, symbols: list):
        """Handle subscribe request."""
        subscribed = []
        
        for symbol in symbols:
            symbol = symbol.upper()
            
            # Add to subscriptions
            if symbol not in self.subscriptions:
                self.subscriptions[symbol] = set()
            self.subscriptions[symbol].add(client)
            
            # Subscribe to IB if needed
            if symbol not in self.tickers:
                if await self._subscribe_symbol(symbol):
                    subscribed.append(symbol)
                    
                    # Wait a moment for initial data
                    await asyncio.sleep(0.5)
                    
                    # Send initial snapshot
                    if symbol in self.tickers:
                        data = self._ticker_to_dict(symbol, self.tickers[symbol])
                        await self._safe_send(client, json.dumps({
                            "type": "price",
                            "symbol": symbol,
                            "data": data
                        }))
            else:
                subscribed.append(symbol)
                # Send current data
                data = self._ticker_to_dict(symbol, self.tickers[symbol])
                await self._safe_send(client, json.dumps({
                    "type": "price",
                    "symbol": symbol,
                    "data": data
                }))
        
        await self._safe_send(client, json.dumps({
            "type": "subscribed",
            "symbols": subscribed
        }))
    
    async def _handle_unsubscribe(self, client: WebSocketServerProtocol, symbols: list):
        """Handle unsubscribe request."""
        unsubscribed = []
        
        for symbol in symbols:
            symbol = symbol.upper()
            
            if symbol in self.subscriptions:
                self.subscriptions[symbol].discard(client)
                unsubscribed.append(symbol)
                
                # If no more clients, unsubscribe from IB
                if not self.subscriptions[symbol]:
                    del self.subscriptions[symbol]
                    await self._unsubscribe_symbol(symbol)
        
        await self._safe_send(client, json.dumps({
            "type": "unsubscribed",
            "symbols": unsubscribed
        }))
    
    async def _handle_snapshot(self, client: WebSocketServerProtocol, symbols: list):
        """Handle snapshot request (one-time quote)."""
        if not self.ib or not self.ib.isConnected():
            await self._safe_send(client, json.dumps({
                "type": "error",
                "message": "IB not connected"
            }))
            return
        
        for symbol in symbols:
            symbol = symbol.upper()
            
            try:
                contract = Stock(symbol, 'SMART', 'USD')
                await self.ib.qualifyContractsAsync(contract)
                
                ticker = self.ib.reqMktData(contract, '', True, False)  # snapshot=True
                await asyncio.sleep(1)
                
                data = self._ticker_to_dict(symbol, ticker)
                await self._safe_send(client, json.dumps({
                    "type": "snapshot",
                    "symbol": symbol,
                    "data": data
                }))
                
            except Exception as e:
                await self._safe_send(client, json.dumps({
                    "type": "error",
                    "message": f"Failed to get snapshot for {symbol}: {e}"
                }))
    
    async def run(self):
        """Run the WebSocket server."""
        if not HAS_WS:
            logger.error("websockets library not installed")
            return
        
        # Connect to IB
        if not await self.connect_ib():
            logger.warning("Starting without IB connection (will retry)")
        
        self.running = True
        
        # Start WebSocket server
        async with websockets.serve(self.handle_client, "0.0.0.0", self.ws_port):
            logger.info(f"✓ WebSocket server running on ws://0.0.0.0:{self.ws_port}")
            
            # Keep running until stopped
            while self.running:
                await asyncio.sleep(1)
                
                # Periodically check IB connection
                if self.ib and not self.ib.isConnected():
                    await self._reconnect_ib()
    
    async def shutdown(self):
        """Gracefully shutdown the server."""
        logger.info("Shutting down...")
        self.running = False
        
        # Disconnect from IB
        if self.ib and self.ib.isConnected():
            self.ib.disconnect()
        
        # Close all client connections
        for client in list(self.clients):
            try:
                await client.close()
            except:
                pass


async def main():
    parser = argparse.ArgumentParser(description="IB Real-Time Price Server")
    parser.add_argument("--port", type=int, default=8765, help="WebSocket server port")
    parser.add_argument("--ib-host", default="127.0.0.1", help="IB Gateway host")
    parser.add_argument("--ib-port", type=int, default=4001, help="IB Gateway port")
    args = parser.parse_args()
    
    server = IBRealtimeServer(
        ws_port=args.port,
        ib_host=args.ib_host,
        ib_port=args.ib_port
    )
    
    # Handle shutdown signals
    loop = asyncio.get_event_loop()
    for sig in (signal.SIGINT, signal.SIGTERM):
        loop.add_signal_handler(sig, lambda: asyncio.create_task(server.shutdown()))
    
    await server.run()


if __name__ == "__main__":
    if not HAS_IB or not HAS_WS:
        print("Missing dependencies. Install with:")
        print("  pip install ib_insync websockets")
        sys.exit(1)
    
    asyncio.run(main())
